core/psh/psh.c

```diff
@@ -972,7 +974,117 @@ static void psh_signalstop(int sig)
 }
 
 
-static int psh_run(void)
+static int psh_login(void)
+{
+	int insz = -1, code;
+	const int maxinsz = 32;
+	char c, indata[maxinsz+1], indatasave[maxinsz+1], shdw[128];
+	struct termios mode;
+
+	tcgetattr(0, &mode);
+	mode.c_lflag |= (ICANON | ECHO);
+	tcsetattr(0, TCSANOW, &mode);
+
+	/* Get login from user */
+	memset(indata, '\0', sizeof(indata));
+
+	for (;;) {
+		if (insz < 0) {
+			fprintf(stdout, "Login: ");
+			fflush(stdout);
+			insz++;
+		}
+
+		if(read(STDIN_FILENO, &c, 1) != 1)
+			continue;
+			
+		if (c == '\n') {
+			if (insz <= maxinsz && insz > 0) {
+				memset(indatasave, '\0', sizeof(indatasave));
+				memcpy(indatasave, indata, sizeof(indatasave));
+				memset(indata, '\0', sizeof(indata));
+				break;
+			}
+			else {
+				memset(indata, '\0', sizeof(indata));
+				insz = -1;
+				continue;
+			}
+		}
+		else if (isprint(c)) {
+			if (insz < maxinsz)
+				indata[insz] = c;
+			insz++;
+			continue;
+		}
+	}
+
+	/* Get password from user */
+	memset(shdw, '\0', sizeof(shdw));
+	memset(indata, '\0', sizeof(indata));
+	mode.c_lflag &= ~(ECHO);
+	tcsetattr(0, TCSANOW, &mode);
+	insz = -1;
+	for (;;) {
```

Comment:
30 lines of duplicated code. Move to subroutine?

Reply:
Moved to subroutine. Code is now more elegant. 

Before the changes there were differences in credential getting loops and i thought that conditionally distinguishing them in subroutine would bloat the code to similar sizes. However after properly using canonical mode it is obsolete.